File: src/Climate.py
import pandas as pd

from .Component import Component


class Climate(Component):

    """Serves as an interface to climate data"""

    def __init__(self, data, **kwargs):
        """
        Parameters
        ----------
        * data: str, path and filename of table-like data
        """
        self.data = pd.read_csv(data, index_col=0, parse_dates=True)

        # Set all key word arguments as attributes
        for key, value in kwargs.items():
            setattr(self, key, value)
        # End For

        climate_year = self.data.groupby(by=self.data.index.year).sum()
        self.description = climate_year.describe()
        self.description.loc['90%', :] = climate_year.quantile(q=0.9)

        # self.min_rainfall = self.description.loc['min', 'rainfall']
        # self.max_rainfall = self.description.loc['max', 'rainfall']
        # self.med_rainfall = self.description.loc['50%', 'rainfall']  # Median rainfall
        # self.mean_rainfall = self.description.loc['mean', 'rainfall']
        # self.high_rainfall = self.description.loc['90%', 'rainfall']

    # End init()

    def __getattr__(self, attr):
        return getattr(self.data, attr)
    # End __getattr__()

    # def get_climate_stat(self, attrib, phenom='rainfall'):
    #     """Retrieve climate statistics.

    #     e.g. standard deviation of rainfall

    #     Equivalent to:
    #     `return self.data.groupby(by=self.data.index.year).sum().describe().loc['std', :]`

    #     Examples
    #     --------
    #     ```python
    #     >>> Climate.get_climate_stat('std', 'rainfall')
    #     >>> Climate.get_climate_stat('mean', 'rainfall')
    #     ```

    #     Parameters
    #     ----------
    #     * attrib : str, Climate attribute accessible through a Pandas Dataframe describe() or 90th percentile
    #     * phenom : str, desired climate phenomenon, e.g. rainfall or ET
    #     """
    #     return self.description.loc[attrib, phenom]
    # # End get_climate_stat()

    def annual_rainfall(self, timestep):
        """
        Calculate the total amount of rainfall that occured in a year, given in the timestep

        Parameters
        ----------
        * timestep : datetime or int, indicating year in terms of time step.
        """
        year = timestep if type(timestep) == int else timestep.year
        data = self.data

        year_data = data[data.index.year == year]
        yearly_rainfall = year_data.sum()

        return yearly_rainfall['rainfall']
    # End annual_rainfall()

    def get_season_range(self, start, end):
        """Gets climate data for season range.

        Parameters
        ----------
        * start : datetime, start of range in Y-m-d format, inclusive.
        * end : datetime, end of range in Y-m-d format, inclusive.
        * data : DataFrame, data to extract data from.
        """
        data = self.data
        mask = (data.index >= start) & (data.index <= end)

        return data.loc[mask]
    # End get_season_range()

    def _ensure_datetime(self, start, end):
        if type(start) == str:
            start = pd.to_datetime(start)
        # End if

        if type(end) == str:
            end = pd.to_datetime(end)
        # End if

        assert end > start, 'Season end date cannot be earlier than start date ({} < {} ?)'.format(start, end)

        return start, end
    # End _ensure_datetime()

    def get_seasonal_rainfall(self, season_range, partial_name):
        """Retrieve seasonal rainfall.

        Parameters
        ----------
        * season_range : List-like, start and end dates, can be string or datetime object
        * data : pd.Series, to slice

        Returns
        --------
        numeric of seasonal rainfall
        """
        start, end = self._ensure_datetime(*season_range)
        rain_cols = [c for c in self.data.columns if ('rainfall' in c) and (partial_name in c)]

        return self.get_season_range(start, end).loc[:, rain_cols].sum()[0]
    # End get_seasonal_rainfall()

    def get_seasonal_et(self, season_range, partial_name):
        """Retrieve seasonal rainfall.

        Parameters
        ----------
        * season_range : List-like, start and end dates, can be string or datetime object

        Returns
        --------
        numeric of seasonal rainfall
        """
        start, end = self._ensure_datetime(*season_range)
        et_cols = [c for c in self.data.columns if ('ET' in c) and (partial_name in c)]

        return self.get_season_range(start, end).loc[:, et_cols].sum()[0]
    # End get_seasonal_et()

# End Climate()

File: src/Climate.jl
using DataFrames, Dates


"""Serves as an interface to climate data"""
mutable struct Climate <: AgComponent

    data::Any
    time_steps::Array
    description::Any
    annual_rainfall::DataFrame

    function Climate(data::DataFrame)
        c = new()
        c.data = data
        c.time_steps = data[!, :Date]
        data[:, :Year] = Dates.year.(c.time_steps)

        gdf = groupby(data[:, filter(x -> x != :Date, propertynames(data))], :Year)

        not_year_col::Array{Symbol,1} = collect(filter(x -> x != :Year, propertynames(gdf)))
        yearly = combine(gdf, names(gdf) .=> sum)

        rain_cols::Array{Symbol,1} = filter(x -> occursin("rainfall", string(x)), propertynames(yearly))
        push!(rain_cols, :Year)

        c.annual_rainfall = yearly[:, rain_cols]

        stats::Array{Symbol,1} = [:mean, :std, :min, :q25, :median, :q75, :max, :eltype, :nunique, :nmissing]
        not_year_col = filter(x -> x != :Year, propertynames(yearly))

        c.description = describe(yearly[:, not_year_col], stats...)

        return c
    end
end

function Base.getindex(c::Climate, args...)::DataFrame
    return c.data[args...]
end

# function __getitem__(c::Climate, item)
#     return c.data[item]
# end

# function annual_rainfall(c::Climate, timestep::Date)
#     """
#     Calculate the total amount of rainfall that occured in a year, given in the timestep

#     Parameters
#     ----------
#     * timestep : datetime or int, indicating year in terms of time step.
#     """
#     year = timestep if type(timestep) == int else timestep.year
#     data = c.data

#     year_data = data[data.index.year == year]
#     yearly_rainfall = year_data.sum()

#     return yearly_rainfall['rainfall']
# end

# function annual_rainfall(c::Climate, timestep::Int64)
#     """
#     Calculate the total amount of rainfall that occured in a year, by index position.

#     Parameters
#     ----------
#     * timestep : Indicating year in terms of index position.
#     """
#     data = c.data

#     year_data = data[data.index.year == timestep]
#     yearly_rainfall = sum(year_data)

#     return yearly_rainfall['rainfall']
# end

"""Gets climate data for season range.

Parameters
----------
* p_start : datetime, start of range in Y-m-d format, inclusive.
* p_end : datetime, end of range in Y-m-d format, inclusive.
"""
function get_season_range(c::Climate, p_start::Date, p_end::Date)::DataFrame
    dates::Array{Date,1} = c.time_steps::Array{Date,1}
    mask::BitArray{1} = (p_start .<= dates .<= p_end)::BitArray{1}
    return c.data[mask, :]
end


"""Converts strings to datetime."""
function ensure_date(c::Climate, p_start::String, p_end::String)::Tuple
    s::Date = Date(p_start)
    e::Date = Date(p_end)

    @assert e > s "Season end date cannot be earlier than start date ($(p_start) < $(p_end) ?)"

    return s, e
end


"""Retrieve seasonal rainfall by matching column name. 
Columns names are expected to have 'rainfall' with some identifier.

Parameters
----------
* season_range : List-like, start and end dates, can be string or datetime object
* partial_name : str, string to (partially) match column name identifier on

Example
----------
Where column names are: 'rainfall_field1', 'rainfall_field2', ...

`get_seasonal_rainfall(c, ['1981-01-01', '1982-06-01'], 'field1')`

Returns
--------
numeric, representing seasonal rainfall
"""
function get_seasonal_rainfall(c::Climate, season_range::Array{Date}, partial_name::String)::Float64
    s::Date, e::Date = season_range
    subset::DataFrame = get_season_range(c, s, e)[!, Regex("rainfall")][!, Regex(partial_name)]
    return sum.(eachcol(subset))[1]
end


"""Retrieve seasonal evapotranspiration.

Parameters
----------
* season_range : List-like, start and end dates, can be string or datetime object
* partial_name : str, string to (partially) match column name identifier on

Returns
--------
numeric of seasonal rainfall
"""
function get_seasonal_et(c::Climate, season_range::Array{Date}, partial_name::String)::DataFrame
    s::Date, e::Date = season_range
    subset::DataFrame = get_season_range(c, s, e)[!, Regex("ET")][!, Regex(partial_name)]
    return sum.(eachcol(subset))[1]
end
